pkg/kubernetes/kubernetes.go

```diff
@@ -0,0 +1,274 @@
+	if context == "" {
+		// Use default context
+		return clientcmd.BuildConfigFromFlags("", kubeconfig)
+	}
+
+	return clientcmd.NewNonInteractiveDeferredLoadingClientConfig(
+		&clientcmd.ClientConfigLoadingRules{ExplicitPath: kubeconfig},
+		&clientcmd.ConfigOverrides{CurrentContext: context}).ClientConfig()
+}
+
+// ResourceID returns a human readable identifier indicating the object kind, name, and namespace.
+func ResourceID(obj runtime.Object) string {
+	m, err := meta.Accessor(obj)
+	if err != nil {
+		return ""
+	}
+
+	gvk := obj.GetObjectKind().GroupVersionKind()
+
+	return fmt.Sprintf("%s:%s/%s", gvk.Kind, m.GetNamespace(), m.GetName())
+}
+
+// Namespaced sets the namespace on an object to namespace, if it is a namespace scoped resource.
+// If the resource is cluster scoped, then it is ignored and the namespace is not set.
+// If it is a namespaced resource and a namespace is already set, then the namespace is unchanged.
+func Namespaced(dClient discovery.DiscoveryInterface, obj runtime.Object, namespace string) (string, string, error) {
+	m, err := meta.Accessor(obj)
+	if err != nil {
+		return "", "", err
+	}
+
+	if m.GetNamespace() != "" {
+		return m.GetName(), m.GetNamespace(), nil
+	}
+
+	resource, err := GetAPIResource(dClient, obj.GetObjectKind().GroupVersionKind())
+	if err != nil {
+		return "", "", fmt.Errorf("retrieving API resource for %v failed: %v", obj.GetObjectKind().GroupVersionKind(), err)
+	}
+
+	if !resource.Namespaced {
+		return m.GetName(), "", nil
+	}
+
+	m.SetNamespace(namespace)
+	return m.GetName(), namespace, nil
+}
+
+// MatchesKind returns true if the Kubernetes kind of obj matches any of kinds.
+func MatchesKind(obj runtime.Object, kinds ...runtime.Object) bool {
+	gvk := obj.GetObjectKind().GroupVersionKind()
+
+	for _, kind := range kinds {
+		if kind.GetObjectKind().GroupVersionKind() == gvk {
+			return true
+		}
+	}
+
+	return false
+}
+
+// ObjectKey returns an instantiated ObjectKey for the provided object.
+func ObjectKey(obj runtime.Object) client.ObjectKey {
+	m, _ := meta.Accessor(obj) //nolint:errcheck // runtime.Object don't have the error issues of interface{}
+	return client.ObjectKey{
+		Name:      m.GetName(),
+		Namespace: m.GetNamespace(),
+	}
+}
+
+// NewV1Pod returns a new corev1.Pod object.
+// Each of name, namespace and serviceAccountName are set if non-empty.
+func NewV1Pod(name, namespace, serviceAccountName string) *corev1.Pod {
+	pod := corev1.Pod{}
+	pod.SetGroupVersionKind(schema.GroupVersionKind{
+		Group:   "",
+		Version: "v1",
+		Kind:    "Pod",
+	})
+	if name != "" {
+		pod.SetName(name)
+	}
+	if namespace != "" {
+		pod.SetNamespace(namespace)
+	}
+	if serviceAccountName != "" {
+		pod.Spec.ServiceAccountName = serviceAccountName
+	}
+	return &pod
+}
+
+// FakeDiscoveryClient returns a fake discovery client that is populated with some types for use in
+// unit tests.
+func FakeDiscoveryClient() discovery.DiscoveryInterface {
```

Comment:
This function seems to be used only by tests. Please move it to a separate package such as `pkg/kubernetes/fake`. It might then make sense to drop the `Fake` prefix from the function name to avoid stuttering.